### PR TITLE
Correct Bayer2Gray u8 SIMD

### PR DESCRIPTION
**Pull request by FantasqueX**
SIMD version of CV_DESCALE is not correct. It should be implemented using v_dotprod.

What's more, the stop condition of vector operation should be `bayer < bayer_end - 14` because we just need to make sure result is safely stored into `dst`.

Closes: https://github.com/opencv/opencv/issues/25823

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by FantasqueX** [comment]
Hi @opencv-alalek May I ask why you add "pr: needs test" label? `Imgproc_ColorBayer.regression` can cover the function I change. Do I still need to add a new test?

**Comment by asmorkalov** [comment]
The test passed before and is green now. It means that it's not representative. Could you add something that highlights the issue?

**Comment by FantasqueX** [comment]
@asmorkalov Hi, I have added a regression test which fails without my patch.

File: modules/imgproc/test/test_color.cpp [hunk 1]
```diff
@@ -1863,6 +1863,31 @@ TEST(Imgproc_ColorBayer, regression)
...
+
+    cvtColor(src, dst, COLOR_BayerBG2GRAY);
+
+    for (int i = 0; i < dst.rows; ++i)
+    {
+        for (int j = 0; j < dst.cols; ++j)
+        {
+            EXPECT_EQ(1, dst.at<uchar>(i, j));
+        }
+    }
```
Comment (asmorkalov) [inline modules/imgproc/test/test_color.cpp:1888]:
```
cv::Mat gt(rows, cols, cv::Scalar(1));
EXPECT_EQ(0, cv::norm(dst, gt, NORM_INF));
```

**Review by asmorkalov — APPROVED** [review]
Looks good for me in general

**Comment by asmorkalov** [comment]
@mshabunin could you take a look too?

**Review by asmorkalov — APPROVED** [review]

**Pull request merged**